bublecs/bublecs.h

```diff
@@ -3,188 +3,309 @@
+        entity.type = GetCompositeType<Components...>();
+
+        AddToArchetypes(entity, std::forward<Components>(components)...);
+
+        entities.emplace(entity.id, entity);
+
+        return entity.id;
+
+    }
+
+    std::pair<ECSId, Archetype&> FindOrCreateArchtype(const CompositeType& type)
+    {
+        auto archetype = std::find_if(archetypes.begin(), archetypes.end(), [&](auto& p) {
+            auto& [id, arch] = p;
+            return  arch.type.Equal(type);
+        });
+
+        if (archetype != archetypes.end()) {
+            auto [id, arch] = *archetype;
+            return {id, arch};
+        }
+
+        Archetype newArchtype;
+        newArchtype.columns.resize(type.Size());
+        newArchtype.type = type;
+        newArchtype.rowsCount = 0;
+        ECSId arhc_id = GenerateNewArchetypeId();
+        archetypes.emplace(arhc_id, newArchtype);
+        return {arhc_id, archetypes[arhc_id]};
+    }
+
+    template <typename... Components>
+    void AddToArchetypes(Entity& entity, Components&&... components)
+    {
+        auto [id, archtype] = FindOrCreateArchtype(entity.type);
+        entity.archetype_id = id;
+        entity.row = archtype.rowsCount;
+        archtype.AddRow(std::forward<std::remove_reference_t<Components>>(components)...);
+    }
+
+    template <typename... Components>
+    void RemoveComponents(ECSId entity_id)
+    {
+        CompositeType type = GetCompositeType<Components...>();
+
+        // if entity found - do remove components
+        if (entities.count(entity_id)) {
+            Entity& entity = entities.at(entity_id);
+            //entities_to_delete.push_back(entity);
+            Archetype arch = archetypes[entity.archetype_id];
+            CompositeType new_cmp_type = arch.type;
+
+            // fill components ids to delete
+            std::unordered_map<ECSId, ComponentType> components_to_delete;
+            ([&] {
+                ECSId type_id = GetTypeId<Components>();
+                components_to_delete.emplace(type_id, arch.type.GetType(type_id));
+                new_cmp_type.RemoveComponent(type_id);
+            }(),
+            ...);
+
+            // copy row from archtype to tmp var without components to delete
+            std::vector<Column> new_columns;
+            for (size_t i = 0; i < arch.columns.size(); ++i ) {
+                ECSId type_id = arch.type.GetTypeIdx(i);
+                auto start = arch.columns[i].begin() + entity.row;
+                auto cmp_size = arch.type.GetTypeSize(type_id);
+
+                bool leave = components_to_delete.count(type_id) == 0; // if not in components to delete
+                if (leave) {
+                    new_columns.push_back({start, start + cmp_size});
+                }
+
+                // move row to delete to the end
+                for (size_t j = 0; j < cmp_size; ++j) {
+                    std::swap(*(start + j), *(arch.columns[i].end() - cmp_size + j));
```

Comment:
If we would call RemoveComponent from a system this could break iteration. As soon as we want to make iteration as lightweight as possible, lets just save entity id in removal list for RemoveComponent and do all that dirty work in GC call.